src/cascadia/WinRTUtils/inc/ThrottledFunc.h

```diff
@@ -54,88 +58,119 @@ class ThrottledFunc : public std::enable_shared_from_this<ThrottledFunc<leading,
     template<typename... MakeArgs>
     void Run(MakeArgs&&... args)
     {
-        if (!_storage.emplace(std::forward<MakeArgs>(args)...))
-        {
-            _leading_edge();
-        }
+        _lead(std::make_tuple(std::forward<MakeArgs>(args)...));
     }
 
-    // Modifies the pending arguments for the next function
-    // invocation, if there is one pending currently.
-    //
-    // `func` will be invoked as func(Args...). Make sure to bind any
-    // arguments in `func` by reference if you'd like to modify them.
     template<typename F>
     void ModifyPending(F func)
     {
-        _storage.modify_pending(func);
+        const auto guard = _lock.lock_exclusive();
+        if (_pendingRunArgs)
+        {
+            std::apply(func, *_pendingRunArgs);
+        }
     }
 
 private:
     static void __stdcall _timer_callback(PTP_CALLBACK_INSTANCE /*instance*/, PVOID context, PTP_TIMER /*timer*/) noexcept
+    try
     {
-        static_cast<ThrottledFunc*>(context)->_trailing_edge();
+        static_cast<ThrottledFunc*>(context)->_trail();
     }
+    CATCH_LOG()
 
-    void _leading_edge()
+    wil::unique_threadpool_timer _create_timer()
     {
-        if constexpr (leading)
+        wil::unique_threadpool_timer timer{ CreateThreadpoolTimer(&_timer_callback, this, nullptr) };
+        THROW_LAST_ERROR_IF(!timer);
+        return timer;
+    }
+
+    void _lead(std::tuple<Args...> args)
+    {
+        bool timerRunning = false;
+
         {
-            _dispatcher.TryEnqueue(winrt::Windows::System::DispatcherQueuePriority::Normal, [weakSelf = this->weak_from_this()]() {
-                if (auto self{ weakSelf.lock() })
-                {
-                    try
-                    {
-                        self->_func();
-                    }
-                    CATCH_LOG();
+            const auto guard = _lock.lock_exclusive();
+
+            timerRunning = _timerRunning;
+            _timerRunning = true;
+
+            if (!timerRunning && _leading)
+            {
+                // Call the function immediately on the leading edge.
+                // See below (out of lock).
+            }
+            else if (_trailing)
+            {
+                _pendingRunArgs.emplace(std::move(args));
+            }
+        }
 
-                    SetThreadpoolTimerEx(self->_timer.get(), &self->_delay, 0, 0);
-                }
-            });
+        const auto execute = !timerRunning && _leading;
+        const auto schedule = !timerRunning || _debounce;
+
+        if (execute)
+        {
+            _execute(std::move(args), schedule);
```

Comment:
doesn't this `move` from `args` twice, once on R106 and once on R115, if the function is both leading and trailing?